File: lib/android/app/src/main/java/com/reactnativenavigation/react/NavigationReactInitializer.java
package com.reactnativenavigation.react;

import com.facebook.react.ReactHost;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.bridge.ReactContext;
import com.reactnativenavigation.NavigationActivity;
import com.reactnativenavigation.react.events.EventEmitter;

import androidx.annotation.NonNull;

public class NavigationReactInitializer implements ReactInstanceManager.ReactInstanceEventListener {

    private final ReactHost reactHost;
    private boolean waitingForAppLaunchEvent = true;
    private boolean isActivityReadyForUi = false;

    NavigationReactInitializer(ReactHost reactHost) {
        this.reactHost = reactHost;
    }

    void onActivityCreated() {
        reactHost.addReactInstanceEventListener(this);
        waitingForAppLaunchEvent = true;
    }

    void onActivityResumed(NavigationActivity activity) {

            reactHost.onHostResume(activity, activity);
            isActivityReadyForUi = true;
            prepareReactApp();
    }

    void onActivityPaused(NavigationActivity activity) {
        isActivityReadyForUi = false;
        reactHost.onHostPause(activity);
    }

    void onActivityDestroyed(NavigationActivity activity) {
        reactHost.removeReactInstanceEventListener(this);
        reactHost.onHostDestroy(activity);
    }

    private void prepareReactApp() {
        reactHost.start();
        if (waitingForAppLaunchEvent) {
            if (reactHost.getCurrentReactContext() != null) {
                emitAppLaunched(reactHost.getCurrentReactContext());
            }
        }
    }

    private void emitAppLaunched(@NonNull ReactContext context) {
        if (!isActivityReadyForUi) return;
        waitingForAppLaunchEvent = false;
        new EventEmitter(context).appLaunched();
    }

    @Override
    public void onReactContextInitialized(final ReactContext context) {
        emitAppLaunched(context);
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/react/NavigationReactInitializer.kt
package com.reactnativenavigation.react

import com.facebook.react.ReactInstanceEventListener
import com.facebook.react.ReactInstanceManager
import com.facebook.react.bridge.ReactContext
import com.reactnativenavigation.NavigationActivity
import com.reactnativenavigation.react.events.EventEmitter

class NavigationReactInitializer internal constructor(
    private val reactInstanceUtils: ReactInstanceUtils,
    isDebug: Boolean
) :
    ReactInstanceEventListener {
    private val devPermissionRequest = DevPermissionRequest(isDebug)
    private var waitingForAppLaunchEvent = true
    private var isActivityReadyForUi = false

    fun onActivityCreated() {
        reactInstanceUtils.addReactInstanceEventListener(this)
        waitingForAppLaunchEvent = true
    }

    fun onActivityResumed(activity: NavigationActivity) {
        if (devPermissionRequest.shouldAskPermission(activity)) {
            devPermissionRequest.askPermission(activity)
        } else {
            reactInstanceUtils.onHostResume(activity, activity)
            isActivityReadyForUi = true
            prepareReactApp()
        }
    }

    fun onActivityPaused(activity: NavigationActivity) {
        isActivityReadyForUi = false
        if (reactInstanceUtils.hasStartedCreatingInitialContext()) {
            reactInstanceUtils.onHostPause(activity)
        }
    }

    fun onActivityDestroyed(activity: NavigationActivity) {
        reactInstanceUtils.removeReactInstanceEventListener(this)
        if (reactInstanceUtils.hasStartedCreatingInitialContext()) {
            reactInstanceUtils.onHostDestroy(activity)
        }
    }

    private fun prepareReactApp() {
        if (shouldCreateContext()) {
            reactInstanceUtils.createReactContextInBackground()
        } else if (waitingForAppLaunchEvent) {
            reactInstanceUtils.getContext()?.let {
                emitAppLaunched(it)
            }
        }
    }

    private fun emitAppLaunched(context: ReactContext) {
        if (!isActivityReadyForUi) return
        waitingForAppLaunchEvent = false
        EventEmitter(context).appLaunched()
    }

    private fun shouldCreateContext(): Boolean {
        return !reactInstanceUtils.hasStartedCreatingInitialContext()
    }

    override fun onReactContextInitialized(context: ReactContext) {
        emitAppLaunched(context)
    }
}